src/code.cloudfoundry.org/gorouter/route/hash_based.go

```diff
@@ -0,0 +1,251 @@
+package route
+
+import (
+	"context"
+	"errors"
+	"log/slog"
+	"sync"
+
+	log "code.cloudfoundry.org/gorouter/logger"
+)
+
+// HashBased load balancing algorithm distributes requests based on a hash of a specific header value.
+// The sticky session cookie has precedence over hash-based routing and the request should be routed to the instance stored in the cookie.
+// If requests do not contain the hash-related header set configured for the hash-based route option, use the default load-balancing algorithm.
+type HashBased struct {
+	lock *sync.Mutex
+
+	logger               *slog.Logger
+	pool                 *EndpointPool
+	lastEndpoint         *Endpoint
+	lastLookupTableIndex uint64
+
+	stickyEndpointID string
+	mustBeSticky     bool
+
+	HeaderValue string
+}
+
+// NewHashBased initializes an endpoint iterator that selects endpoints based on a hash of a header value.
+// The global properties locallyOptimistic and localAvailabilityZone will be ignored when using Hash-Based Routing.
+func NewHashBased(logger *slog.Logger, p *EndpointPool, initial string, mustBeSticky bool, headerValue string) EndpointIterator {
+	return &HashBased{
+		logger:           logger,
+		pool:             p,
+		lock:             &sync.Mutex{},
+		stickyEndpointID: initial,
+		mustBeSticky:     mustBeSticky,
+		HeaderValue:      headerValue,
+	}
+}
+
+// Next selects the next endpoint based on the hash of the header value.
+// If a sticky session endpoint is available and not overloaded, it will be returned.
+// If the request must be sticky and the sticky endpoint is unavailable or overloaded, nil will be returned.
+// If no sticky session is present, the endpoint will be selected based on the hash of the header value.
+// It returns the same endpoint for the same header value consistently.
+// If the hash lookup fails or the endpoint is not found, nil will be returned.
+func (h *HashBased) Next(attempt int) *Endpoint {
+	h.lock.Lock()
+	defer h.lock.Unlock()
+
+	endpoint := h.findEndpointIfStickySession()
+	if endpoint == nil && h.mustBeSticky {
+		return nil
+	}
```

Comment:
can this not be turned around? if `h.mustBeSticky` is false we can skip the whole `findEndpointIfStickySession()` call?

The rest of the non-sticky handling remains as is.

Reply:
findEndpointIfStickySession() is intended to be the same as for another EndpointIterators. It is not only about h.mustBeStricky, but also for the case where the initialEndpoint (sticky endpoint) is set.